source/Halibut/Transport/SecureListener.cs

```diff
@@ -246,18 +250,28 @@ async Task ExecuteRequest(TcpClient client)
 
                     if (Authorize(thumbprint, clientName))
                     {
-                        // The ExchangeMessage call can hang on reading the stream which keeps a thread alive,
-                        // so we dispose the stream which will cause the thread to abort with an exceptions.
-                        var weakSsl = new WeakReference(ssl);
+                        if (asyncHalibutFeature == AsyncHalibutFeature.Disabled)
+                        {
+                            // The ExchangeMessage call can hang on reading the stream which keeps a thread alive,
+                            // so we dispose the stream which will cause the thread to abort with an exceptions.
+                            var weakSsl = new WeakReference(ssl);
 #if !NETFRAMEWORK
-                        await
+                            await
 #endif
-                        using (cts.Token.Register(() =>
-                                     {
-                                         if (weakSsl.IsAlive)
-                                             ((IDisposable)weakSsl.Target).Dispose();
-                                     }))
+                            using (cancellationToken.Register(() =>
+                                   {
+                                       if (weakSsl.IsAlive)
+                                           ((IDisposable)weakSsl.Target).Dispose();
+                                   }))
+                            {
+                                tcpClientManager.AddActiveClient(thumbprint, client);
+                                await ExchangeMessages(ssl).ConfigureAwait(false);
+                            }
+                        }
+                        else
                         {
+                            // The stream is wrapped in a NetworkTimeoutStream which handles closing the inner stream on timeout
```

Comment:
For Async Halibut we don't need this weak ref and disposal as we wrap all Async Halibut calls so they cancel / timeout / close / dispose

Reply:
@LukeButters `TcpClientManager` is now Disposable and the `SecureListener` will Dispose of it

Reply:
Nice thanks! I think the result is simpler to follow code.